### PR TITLE
Replace miscellaneous pyutilib.misc items

### PR DESCRIPTION
**Pull request by mrmundt**
## Summary/Motivation:
As part of the Pyomo 6.0 release, we are separating PyUtilib from Pyomo. This PR replaces or removes items from the `pyutilib.misc` module.

## Changes proposed in this PR:
- Port over `StreamIndenter`
- Add `capture_output` to `pyomo.common.tee`
- Replace various PyUtilib `flatten` methods with a universal `flattener` (which is a wrapper around `normalize_index`)

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/solvers/plugins/solvers/XPRESS.py [hunk 1]
```diff
@@ -303,7 +306,8 @@ def process_logfile(self):
                 results.solver.termination_message = ' '.join(tokens)
 
         try:
-            results.solver.termination_message = yaml_fix(results.solver.termination_message)
+            if isinstance(results.solver.termination_message, basestring):
```
Comment (jsiirola) [inline pyomo/solvers/plugins/solvers/XPRESS.py:309]: This could just be `six.string_types`, right?

**Review by jsiirola — CHANGES_REQUESTED** [review]

File: pyomo/opt/plugins/colin_xml_io.py [hunk 2]
```diff
@@ -79,6 +79,6 @@ def _process(self, response):
             elif type(response[key]) in intlist:
                 text_elt = doc.createTextNode( str(response[key]) )
             else:
-                text_elt = doc.createTextNode( tostr(response[key]) )
+                text_elt = doc.createTextNode(' '.join(repr(x) for x in response[key]))
```
[the remark this answers is not in the document]
Comment (mrmundt) [reply pyomo/opt/plugins/colin_xml_io.py:82]: The original implementation of `tostr` is:
```
def tostr(array):
    """ Create a string from an array of numbers """
    tmpstr = ""
    for val in array:
        tmpstr = tmpstr + " " + repr(val)
    return tmpstr.strip()
```
It uses `repr`, not `str`, so I retained that choice.

File: pyomo/solvers/plugins/solvers/XPRESS.py (hunk 1 above)
Comment (mrmundt) [reply pyomo/solvers/plugins/solvers/XPRESS.py:309]: It, in fact, could be. There's a lot of mix-and-match through Pyomo about whether to use `str`, `basestring`, or `string_types`. We may want to decide and standardize.

Comment (mrmundt) [reply pyomo/solvers/plugins/solvers/XPRESS.py]: I have opened #1823 and assigned it to myself.

File: pyomo/opt/plugins/colin_xml_io.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/opt/plugins/colin_xml_io.py:82]: Fair enough.  That was my guess, too.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**